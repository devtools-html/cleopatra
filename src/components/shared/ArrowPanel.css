/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.arrowPanelAnchor {
  position: absolute;
  z-index: 10;
  left: 50%;
  top: 75%;
}

.arrowPanel {
  --internal-offset-from-right: var(--offset-from-right, 60px);
  --internal-offset-from-top: 15px;
  --internal-width: var(--width, initial);
  --internal-button-height: 30px;
  --internal-ok-button-background: var(--ok-button-background, #7990c8);
  --internal-ok-button-color: var(--ok-button-color, white);
  position: absolute;
  top: var(--internal-offset-from-top);
  right: calc(0px - var(--internal-offset-from-right));
  min-width: var(--internal-width);
  filter: drop-shadow(0 0 0.5px rgba(0,0,0,0.4)) drop-shadow(0 4px 5px rgba(0,0,0,0.4));
  filter: url(../../../res/img/svg/shadowfilter.svg#menushadow);
  color: black;
  text-align: left;
  line-height: 1.3;
  background: hsla(0, 0%, 97%, 0.95);
  border-radius: 5px;
  transform-origin: calc(100% - var(--internal-offset-from-right))
                    calc(0px - var(--internal-offset-from-top));
}

.arrowPanel:not(.open) {
  pointer-events: none;
  opacity: 0;
  /* visibility: hidden hides the panel from the accessibility tree when it's closed */
  visibility: hidden;
  /* The visibility is set to hidden after the panel closing transition. */
  transition: opacity 200ms ease-out, visibility 0s 200ms;
}

.arrowPanel.open {
  transition: none;
  animation: arrowPanelAppear 400ms;
  opacity: 1;
}

@keyframes arrowPanelAppear {
  from {
    opacity: 0;
    transform: scale(0);
    animation-timing-function: ease-out;
  }

  40% {
    opacity: 0.6;
    transform: scale(1);
    animation-timing-function: ease-in-out;
  }

  60% {
    opacity: 0.9;
    transform: scale(0.96);
    animation-timing-function: cubic-bezier(.3,0,0.3,1);
  }
}

.arrowPanelArrow {
  position: absolute;
  left: 0;
  right: 0;
  top: calc(0px - var(--internal-offset-from-top));
  height: var(--internal-offset-from-top);
  overflow: hidden;
}

.arrowPanelArrow::before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: calc(100% - var(--internal-offset-from-right));
  width: calc(1.42 * var(--internal-offset-from-top));
  height: calc(1.42 * var(--internal-offset-from-top));
  background: hsla(0, 0%, 97%, 0.95);
  transform: rotate(45deg);
  transform-origin: top left;
  z-index: -1;
}

.arrowPanelTitle,
.arrowPanelSubTitle {
  font-size: 1.2em;
  padding: 10px;
  margin: 0;
}

.arrowPanelSubTitle {
  border-bottom: 1px solid rgba(0,0,0,0.05);
  padding: 5px 0;
}

.arrowPanelContent {
  padding: 0 10px;
  overflow: auto;
  --internal-approx-distance-from-top: 60px;
  --internal-approx-distance-to-bottom: 100px;
  max-height: calc(100vh - var(--internal-approx-distance-from-top) - var(--internal-button-height) - var(--internal-approx-distance-to-top));
}

.arrowPanel.hasTitle > .arrowPanelContent {
  border-top: 1px solid rgba(0,0,0,0.05);
}

.arrowPanel.hasButtons > .arrowPanelContent {
  border-bottom: 1px solid rgba(0,0,0,0.05);
}

.arrowPanel ul {
  padding-left: 2em;
}

.arrowPanelButtons {
  height: 35px;
  line-height: 35px;
  display: flex;
  flex-flow: row nowrap;
  align-items: stretch;
}

.arrowPanelOkButton,
.arrowPanelCancelButton {
  border: 0;
  background: none;
  padding: 0;
  margin: 0;
  flex: 1;
  text-align: center;
  font-size: 1.2em;
}

.arrowPanelOkButton {
  border-bottom-right-radius: 5px;
  background: var(--internal-ok-button-background);
  color: var(--internal-ok-button-color);
}

.arrowPanelCancelButton {
  border-bottom-left-radius: 5px;
}

.arrowPanelOkButton:hover,
.arrowPanelCancelButton:hover {
  background-image: linear-gradient(rgba(0,0,0,0.1), rgba(0,0,0,0.1));
}

.arrowPanelOkButton:hover:active,
.arrowPanelCancelButton:hover:active {
  background-image: linear-gradient(rgba(0,0,0,0.2), rgba(0,0,0,0.2));
}

.arrowPanelOpenMetaInfo {
  width: 350px;
  font-size: 12px;
  line-height: 20px;
  padding-top: 5px;
  max-height: calc(100vh - 60px);
  overflow-y: auto;
}

.arrowPanelOpenMetaInfo,
.arrowPanelOpenMetaInfo .arrowPanelArrow::before {
  background-color: white;
}

.arrowPanelSection {
  margin-top: 7px;
  margin-bottom: 12px;
}

.metaInfoLabel {
  width: 120px;
  color: var(--grey-40);
  display: inline-block;
}

.metaInfoList {
  margin-top: -19px;
}

.metaInfoList .metaInfoListItem {
  margin-left: 110px;
  list-style: circle;
}
