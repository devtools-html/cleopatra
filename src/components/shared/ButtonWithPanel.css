/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.buttonWithPanel {
  position: relative;
  display: flex;
  flex-flow: column nowrap;
  align-items: stretch;
}

/**
 * This wrapper uses overflow: hidden to clip the actual button.
 * This clip is needed for the vertical slide animation that the
 * profile sharing composite button does.
 * The clip is on .buttonWithPanelButtonWrapper and not on
 * .buttonWithPanel because the latter would clip the panel. */
.buttonWithPanelButtonWrapper {
  display: flex;
  overflow: hidden;
  height: 100%;
  flex-flow: column nowrap;
  align-items: stretch;
}

.buttonWithPanelButtonWrapper:focus-within {
  overflow: visible;
}

.buttonWithPanelButton {
  height: 100%;
  padding: 0 10px;
  margin: 0;
  border: 0;
  background: none;
  font-size: 100%;
}

.buttonWithPanelButton:hover {
  background-image: linear-gradient(rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.1));
}

.buttonWithPanel.open > * > .buttonWithPanelButton,
.buttonWithPanelButton:hover:active {
  background-image: linear-gradient(rgba(0, 0, 0, 0.2), rgba(0, 0, 0, 0.2));
}
