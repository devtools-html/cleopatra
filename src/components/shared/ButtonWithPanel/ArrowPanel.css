/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * .arrowPanel is a wrapper which renders nothing by itself and
 * just establishes the bounds of the panel. The actual panel appearance is
 * rendered by .arrowPanelInnerWrapper.
 * The bottom edge of .arrowPanel is always at the bottom of the viewport.
 * This allows .arrowPanelInnerWrapper to grow as needed and then to stop
 * growing (and become scrollable) once it hits the bottom.
 *
 * The following properties are set by inline style: left, top, width
 * The following CSS variables may be set by users: --arrow-panel-background, --arrow-panel-padding
 */
.arrowPanel {
  --internal-arrow-clip-height: 15px;
  --internal-background: var(--arrow-panel-background, hsl(0deg 0% 97%));
  --internal-padding: var(--arrow-panel-padding, 16px);

  position: fixed;
  z-index: 2;
  bottom: 5px; /* leave a small gap at the bottom of the viewport */
  display: flex;
  flex-flow: column nowrap;
  padding-top: var(--internal-arrow-clip-height);
  margin-bottom: auto;
  color: black;
  line-height: 1.3;
  pointer-events: none;
  text-align: left;
}

.arrowPanelInnerWrapper {
  display: flex;
  min-height: 0;
  flex-flow: column nowrap;
  border: 0.5px solid rgb(0 0 0 / 0.25);
  border-radius: 5px;
  background: var(--internal-background);
  background-clip: padding-box;
  box-shadow: 0 8px 12px rgb(0 0 0 / 0.35);
  pointer-events: initial;
}

.arrowPanelInnerWrapper:not(.open) {
  opacity: 0;
  pointer-events: none;

  /* The visibility is set to hidden after the panel closing transition. */
  transition:
    opacity 200ms ease-out,
    visibility 0s 200ms;

  /* visibility: hidden hides the panel from the accessibility tree when it's closed */
  visibility: hidden;
}

.arrowPanelInnerWrapper.open {
  animation: arrowPanelAppear 200ms cubic-bezier(0.07, 0.95, 0, 1);
}

@media (prefers-reduced-motion) {
  .arrowPanelInnerWrapper.open {
    animation: none;
    opacity: 1;
  }
}

@keyframes arrowPanelAppear {
  from {
    opacity: 0;
    transform: translateY(-25px);
  }

  to {
    opacity: 1;
    transform: none;
  }
}

/**
 * This element clips the arrow to its triangle shape.
 */
.arrowPanelArrowContainer {
  position: relative;
  overflow: hidden;
  height: var(--internal-arrow-clip-height);
  flex-shrink: 0;
  margin-top: calc(-1 * var(--internal-arrow-clip-height));
}

/**
 * This element is the actual arrow, as a (clipped) rotated square.
 *
 * The `left` property is set by inline style, to the offset at which
 * the arrow tip should be located.
 */
.arrowPanelArrow {
  position: absolute;
  top: 0;
  display: block;
  width: calc(1.42 * var(--internal-arrow-clip-height));
  height: calc(1.42 * var(--internal-arrow-clip-height));
  border: 0.5px solid rgb(0 0 0 / 0.25);
  background: var(--internal-background);
  background-clip: padding-box;
  transform: rotate(45deg);
  transform-origin: top left; /* rotate around the arrow tip */
}

/**
 * This element supplies the padding and optional scrolling for the panel contents.
 */
.arrowPanelContent {
  display: flex;
  overflow: auto;
  min-height: 0;
  flex-flow: column nowrap;
  flex-shrink: 1;
  align-items: stretch;
  padding: var(--internal-padding);
}
