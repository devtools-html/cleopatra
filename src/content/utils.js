// @flow
type ValidWeakMapKey = Object | [];

/**
 * Memoize a function's results using a WeakMap and its first parameter.
 */
export function weakMemoize(fn: (ValidWeakMapKey, ...any) => any) {
  const memo = new WeakMap();
  return function (key: ValidWeakMapKey) {
    const cachedValue = memo.get(key);
    if (cachedValue === undefined) {
      const value = fn.apply(null, arguments);
      memo.set(key, value);
      return value;
    }
    return cachedValue;
  };
}
