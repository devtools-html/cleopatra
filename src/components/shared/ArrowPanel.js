/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';

require('./ArrowPanel.css');

type Props = {|
  onOpen?: () => mixed,
  onClose?: () => mixed,
  onOkButtonClick?: () => mixed,
  onCancelButtonClick?: () => mixed,
  className: string,
  children: React.Node,
  title?: string,
  okButtonText?: string,
  cancelButtonText?: string,
|};

type State = {|
  open: boolean,
  isClosing: boolean,
  openGeneration: number,
|};

class ArrowPanel extends React.PureComponent<Props, State> {
  _panelElement: HTMLElement | null = null;
  state = {
    open: false,
    isClosing: false,
    // The open generation is mistakenly being tagged here as being unused.
    // eslint-disable-next-line react/no-unused-state
    openGeneration: 0,
  };

  _takePanelElementRef = (elem: HTMLElement | null) => {
    this._panelElement = elem;
  };

  open() {
    if (this.state.open) {
      return;
    }

    this.setState({ open: true });
    if (this.props.onOpen) {
      this.props.onOpen();
    }
    window.addEventListener('mousedown', this._windowMouseDownListener, true);
  }

  close() {
    this.setState(state => {
      if (!state.open) {
        return null;
      }
      const openGeneration = state.openGeneration + 1;

      setTimeout(this._onCloseAnimationFinish(openGeneration), 400);

      if (this.props.onClose) {
        this.props.onClose();
      }

      window.removeEventListener(
        'mousedown',
        this._windowMouseDownListener,
        true
      );

      return { open: false, isClosing: true, openGeneration };
    });
  }

  _onCloseAnimationFinish(openGeneration: number) {
    return () => {
      this.setState(state => {
        if (state.openGeneration === openGeneration) {
          return { isClosing: false };
        }
        return null;
      });
    };
  }

  componentWillUnmount() {
    window.removeEventListener(
      'mousedown',
      this._windowMouseDownListener,
      true
    );
  }

  _windowMouseDownListener = (e: MouseEvent) => {
    const target: Node = (e.target: any); // make flow happy
    if (
      this.state.open &&
      this._panelElement &&
      !this._panelElement.contains(target)
    ) {
      this.close();
    }
  };

  _onOkButtonClick = () => {
    this.close();
    if (this.props.onOkButtonClick) {
      this.props.onOkButtonClick();
    }
  };

  _onCancelButtonClick = () => {
    this.close();
    if (this.props.onCancelButtonClick) {
      this.props.onCancelButtonClick();
    }
  };

  render() {
    const {
      className,
      children,
      title,
      okButtonText,
      cancelButtonText,
    } = this.props;
    const hasTitle = title !== undefined;
    const hasButtons = okButtonText || cancelButtonText;
    const { open, isClosing } = this.state;
    return (
      <div className="arrowPanelAnchor">
        <div
          className={classNames(
            'arrowPanel',
            { open, hasTitle, hasButtons },
            className
          )}
          ref={this._takePanelElementRef}
        >
          <div className="arrowPanelArrow" />
          {hasTitle ? <h1 className="arrowPanelTitle">{title}</h1> : null}
          {open || isClosing ? (
            <div className="arrowPanelContent">{children}</div>
          ) : null}
          {hasButtons ? (
            <div className="arrowPanelButtons">
              <input
                type="button"
                className="arrowPanelCancelButton"
                value={cancelButtonText}
                onClick={this._onCancelButtonClick}
              />
              <input
                type="button"
                className="arrowPanelOkButton"
                value={okButtonText}
                onClick={this._onOkButtonClick}
              />
            </div>
          ) : null}
        </div>
      </div>
    );
  }
}

export default ArrowPanel;
