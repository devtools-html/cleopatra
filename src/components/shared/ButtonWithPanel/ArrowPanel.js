/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

// This implements a panel with a "arrow" triangle graphic that points to the
// button that triggers it.
// Please do not use this component directly. This is used by ButtonWithPanel
// to wrap the panel content.

import * as React from 'react';
import classNames from 'classnames';

import './ArrowPanel.css';

import type { CssPixels } from 'firefox-profiler/types';

type Props = {|
  +onOpen: () => mixed,
  +onClose: () => mixed,
  +className?: string,
  +children: React.Node,
|};

type State = {|
  +open: boolean,
  +isClosing: boolean,
  +openGeneration: number,
  +panelRect: Rect,
  +arrowOffset: CssPixels,
|};

type Rect = {|
  left: CssPixels,
  top: CssPixels,
  width: CssPixels,
  height: CssPixels,
|};

// The anchor position describes where the arrow panel's arrow should be located
// *in relation to the panel*. Then the panel will be positioned so that the arrow
// points at the center of the anchor element.
//
//                                   +---------------------------+
//                                   |       anchorElement       |
//                                   +---------------------------+
//                                                 ^
//                                                 |
//          +--------------------------------------+--------+
//          |                                               |
//          |                  The panel                    |
//          |                                               |
//          +-----------------------------------------------+
//
//          |<---------------- panelWidth ----------------->|
//                                                  |<----->|
//                                               distanceFromEdge
export type AnchorPosition = {|
  // If set to "right", the arrow will be positioned `distanceFromEdge` pixels
  // from the right edge of the panel.
  // If set to "left", the arrow will be positioned `distanceFromEdge` pixels
  // from the left edge of the panel.
  anchorEdge: 'left' | 'right',
  distanceFromEdge: CssPixels,
|};

export class ArrowPanel extends React.PureComponent<Props, State> {
  closeTimeout = null;
  state = {
    open: false,
    isClosing: false,
    openGeneration: 0,
    panelRect: { left: 0, top: 0, width: 0, height: 0 },
    arrowOffset: 40,
  };

  /* These 2 methods are called from other components.
  /* See https://github.com/firefox-devtools/profiler/issues/1888 and
   * https://github.com/firefox-devtools/profiler/issues/1641 */
  /* eslint-disable-next-line react/no-unused-class-component-methods */
  open(
    anchorElement: HTMLElement,
    panelWidthInCssPx: number,
    anchorPosition: AnchorPosition
  ) {
    if (this.state.open) {
      return;
    }

    const win = anchorElement.ownerDocument.defaultView;
    const anchorRectRelativeToPage = anchorElement.getBoundingClientRect();
    const anchorRectRelativeToViewport = {
      left: anchorRectRelativeToPage.left - win.scrollX,
      top: anchorRectRelativeToPage.top - win.scrollY,
      width: anchorRectRelativeToPage.width,
      height: anchorRectRelativeToPage.height,
    };

    // Compute the location that the arrow panel should point at with its arrow.
    // This location is in the middle of the anchorRect, 75% of the way down.
    const anchorPoint = {
      left:
        anchorRectRelativeToViewport.left +
        anchorRectRelativeToViewport.width * 0.5,
      top:
        anchorRectRelativeToViewport.top +
        anchorRectRelativeToViewport.height * 0.75,
    };

    // Constrain the panel width to the viewport.
    const adjustedPanelWidth = Math.min(panelWidthInCssPx, win.innerWidth);

    // Compute the panel position.
    const preferredPanelLeft = (function () {
      if (anchorPosition.anchorEdge === 'left') {
        return anchorPoint.left - anchorPosition.distanceFromEdge;
      }
      const right = anchorPoint.left + anchorPosition.distanceFromEdge;
      return right - adjustedPanelWidth;
    })();

    // Constrain the panel position so that the panel is fully onscreen.
    const panelLeft = Math.max(
      0,
      Math.min(preferredPanelLeft, win.innerWidth - adjustedPanelWidth)
    );

    this.setState({
      open: true,
      panelRect: {
        left: panelLeft,
        top: anchorPoint.top,
        width: adjustedPanelWidth,
        height: 0, // ignored
      },
      arrowOffset: anchorPoint.left - panelLeft,
    });
  }

  /* eslint-disable-next-line react/no-unused-class-component-methods */
  close() {
    this.setState((state) => {
      if (!state.open) {
        return null;
      }
      const openGeneration = state.openGeneration + 1;

      clearTimeout(this.closeTimeout);
      this.closeTimeout = setTimeout(
        this._onCloseAnimationFinish(openGeneration),
        400
      );

      return { open: false, isClosing: true, openGeneration };
    });
  }

  _onCloseAnimationFinish(openGeneration: number) {
    return () => {
      this.setState((state) => {
        if (state.openGeneration === openGeneration) {
          return { isClosing: false };
        }
        return null;
      });
    };
  }

  _onArrowPanelClick = (e: { target: HTMLElement } & SyntheticMouseEvent<>) => {
    // The arrow panel element contains the element that has the top arrow,
    // that is visually outside the panel. We still want to hide the panel
    // when clicking in this area.
    if (e.target.className !== 'arrowPanelArrow') {
      // Stop the click propagation to reach the _onWindowClick event when the
      // click is visually inside the panel.
      e.stopPropagation();
    }
  };

  // We're calling open and close callbacks in componentDidUpdate because they
  // often run side-effects, so we want them out of the render phase.
  componentDidUpdate(prevProps: Props, prevState: State) {
    if (!prevState.open && this.state.open) {
      // Opening
      this.props.onOpen();
    }

    if (!this.state.open && prevState.isClosing && !this.state.isClosing) {
      // Closing... but only after the animation.
      this.props.onClose();
    }
  }

  componentWillUnmount() {
    clearTimeout(this.closeTimeout);
  }

  render() {
    const { className, children } = this.props;
    const { open, isClosing, panelRect, arrowOffset } = this.state;

    if (!open && !isClosing) {
      return null;
    }

    return (
      <div
        className={classNames('arrowPanel', { open }, className)}
        onClick={this._onArrowPanelClick}
        style={{
          left: `${panelRect.left}px`,
          top: `${panelRect.top}px`,
          width: `${panelRect.width}px`,
        }}
      >
        <div className={classNames('arrowPanelInnerWrapper', { open })}>
          <div className="arrowPanelArrowContainer">
            <div
              className="arrowPanelArrow"
              style={{ left: `${arrowOffset}px` }}
            />
          </div>
          <div className="arrowPanelContent">{children}</div>
        </div>
      </div>
    );
  }
}
