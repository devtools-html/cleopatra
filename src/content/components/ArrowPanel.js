import React, { PureComponent, PropTypes } from 'react';
import classNames from 'classnames';

require('./ArrowPanel.css');

class ArrowPanel extends PureComponent {
  constructor(props) {
    super(props);
    this.state = { open: false };
    this._windowMouseDownListener = this._windowMouseDownListener.bind(this);
    this._panelElementCreated = elem => { this._panelElement = elem; };
    this._onOkButtonClick = this._onOkButtonClick.bind(this);
    this._onCancelButtonClick = this._onCancelButtonClick.bind(this);
  }

  open() {
    if (this.state.open) {
      return;
    }

    this.setState({ open: true });
    if (this.props.onOpen) {
      this.props.onOpen();
    }
    window.addEventListener('mousedown', this._windowMouseDownListener, true);
  }

  close() {
    if (!this.state.open) {
      return;
    }

    this.setState({ open: false });
    if (this.props.onClose) {
      this.props.onClose();
    }
    window.removeEventListener('mousedown', this._windowMouseDownListener, true);
  }

  componentWillUnmount() {
    window.removeEventListener('mousedown', this._windowMouseDownListener, true);
  }

  _windowMouseDownListener(e) {
    if (this.state.open && this._panelElement &&
        !this._panelElement.contains(e.target)) {
      this.close();
    }
  }

  _onOkButtonClick() {
    this.close();
    if (this.props.onOkButtonClick) {
      this.props.onOkButtonClick();
    }
  }

  _onCancelButtonClick() {
    this.close();
    if (this.props.onCancelButtonClick) {
      this.props.onCancelButtonClick();
    }
  }

  render() {
    const { className, children, title, okButtonText, cancelButtonText } = this.props;
    const hasTitle = (title !== undefined);
    const hasButtons = (okButtonText || cancelButtonText);
    const { open } = this.state;
    return (
      <div className='arrowPanelAnchor'>
        <div className={classNames('arrowPanel', { open, hasTitle, hasButtons }, className)} ref={this._panelElementCreated}>
          <div className='arrowPanelArrow'/>
          { hasTitle ? <h1 className='arrowPanelTitle'>{title}</h1> : null }
          <div className='arrowPanelContent'>
            {children}
          </div>
          {
            hasButtons ? (
              <div className='arrowPanelButtons'>
                <input type='button'
                       className='arrowPanelCancelButton'
                       value={cancelButtonText}
                       onClick={this._onCancelButtonClick}/>
                <input type='button'
                       className='arrowPanelOkButton'
                       value={okButtonText}
                       onClick={this._onOkButtonClick}/>
              </div>) : null
          }
        </div>
      </div>
    );
  }
}

ArrowPanel.propTypes = {
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  onOkButtonClick: PropTypes.func,
  onCancelButtonClick: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.any,
  title: PropTypes.string,
  okButtonText: PropTypes.string,
  cancelButtonText: PropTypes.string,
};

export default ArrowPanel;
