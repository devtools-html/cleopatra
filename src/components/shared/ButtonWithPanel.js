/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import classNames from 'classnames';

require('./ButtonWithPanel.css');

type PanelProps = {
  onOpen?: () => mixed,
  onClose?: () => mixed,
};

interface Panel {
  open(): mixed;
  close(): mixed;
}

/**
 * Note about the `panel` prop: we accept any React element whose Component
 * class implements the `Panel` interface above, and has at least the props from
 * `PanelProps` above, and any State type. */
type Props = {
  className: string,
  label: string,
  panel: React.Element<
    Class<Panel & React.Component<$Subtype<PanelProps>, any>>
  >,
  open?: boolean,
  // This prop tells the panel to be open by default, but the open/close state is fully
  // managed by the ButtonWithPanel component.
  defaultOpen?: boolean,
};

type State = {|
  open: boolean,
|};

class ButtonWithPanel extends React.PureComponent<Props, State> {
  _panel: Panel | null = null;

  constructor(props: Props) {
    super(props);
    this.state = { open: !!props.open };
  }

  componentDidMount() {
    // the panel can be closed by pressing the Esc key
    window.addEventListener('keydown', this._onKeyDown);
    if (this.props.open || this.props.defaultOpen) {
      this.openPanel();
    }
  }

  componentWillUnmount() {
    window.removeEventListener('keydown', this._onKeyDown);
  }

  componentWillReceiveProps(props: Props) {
    if (props.open !== this.props.open) {
      this.setState({ open: !!props.open });
    }
  }

  _onPanelOpen = () => {
    this.setState({ open: true });
    if (this.props.panel.props.onOpen) {
      this.props.panel.props.onOpen();
    }
  };

  _onPanelClose = () => {
    this.setState({ open: false });
    if (this.props.panel.props.onClose) {
      this.props.panel.props.onClose();
    }
  };

  _takePanelRef = (panel: Panel | null) => {
    this._panel = panel;
  };

  openPanel() {
    if (this._panel) {
      this._panel.open();
    }
  }
  closePanel() {
    if (this._panel && this.state.open) {
      this._panel.close();
    }
  }

  _onButtonClick = () => {
    this.openPanel();
  };

  _onKeyDown = (e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      this.closePanel();
    }
  };

  render() {
    const { className, label, panel } = this.props;
    const { open } = this.state;
    return (
      <div className={classNames('buttonWithPanel', className, { open })}>
        <div className="buttonWithPanelButtonWrapper">
          <input
            type="button"
            className={classNames(
              'buttonWithPanelButton',
              `${className}Button`
            )}
            value={label}
            onClick={this._onButtonClick}
          />
        </div>
        {React.cloneElement(panel, {
          ref: this._takePanelRef,
          onOpen: this._onPanelOpen,
          onClose: this._onPanelClose,
        })}
      </div>
    );
  }
}

export default ButtonWithPanel;
