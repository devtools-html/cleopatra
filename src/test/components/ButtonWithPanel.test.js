/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { render, fireEvent } from 'react-testing-library';
import ButtonWithPanel from '../../components/shared/ButtonWithPanel';
import ArrowPanel from '../../components/shared/ArrowPanel';
import { ensureExists } from '../../utils/flow';

describe('shared/ButtonWithPanel', () => {
  // renders the button in its default state
  function setup() {
    return render(
      <ButtonWithPanel
        className="button"
        label="My Button"
        panel={
          <ArrowPanel className="panel">
            <div>Panel content</div>
          </ArrowPanel>
        }
      />
    );
  }

  it('renders the ButtonWithPanel with a closed panel', () => {
    const { container } = setup();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a button with a panel', () => {
    const { container } = render(
      <ButtonWithPanel
        className="button"
        label="My Button"
        open={true}
        panel={
          <ArrowPanel className="panel">
            <div>Panel content</div>
          </ArrowPanel>
        }
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled button', () => {
    const { container } = render(
      <ButtonWithPanel
        className="button"
        label="My Button"
        disabled={true}
        panel={
          <ArrowPanel className="panel">
            <div>Panel content</div>
          </ArrowPanel>
        }
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('opens the panel when the button is clicked and closes the panel when the escape key is pressed', () => {
    const { getByValue, container } = setup();

    fireEvent.click(getByValue('My Button'));
    expect(container.firstChild).toMatchSnapshot();

    //it closes the panel when Esc key is pressed
    fireEvent.keyDown(container, {
      key: 'Escape',
      keyCode: 27,
      which: 27,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('opens the panel when the button is clicked and closes the panel by clicking outside the panel', () => {
    const { getByValue, container } = setup();

    fireEvent.click(getByValue('My Button'));
    expect(container.firstChild).toMatchSnapshot();

    //it closes the panel when clicking outside the panel
    const newDiv = ensureExists(document.body).appendChild(
      document.createElement('div')
    );
    fireEvent.mouseDown(newDiv);

    expect(container.firstChild).toMatchSnapshot();
  });
});
