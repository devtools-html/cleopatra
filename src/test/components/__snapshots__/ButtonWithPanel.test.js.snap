// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel by clicking outside the panel 1`] = `
<div
  class="buttonWithPanel button open"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel by clicking outside the panel 2`] = `
<div
  class="buttonWithPanel button"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel when the escape key is pressed 1`] = `
<div
  class="buttonWithPanel button open"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel opens the panel when the button is clicked and closes the panel when the escape key is pressed 2`] = `
<div
  class="buttonWithPanel button"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel renders a button with a panel 1`] = `
<div
  class="buttonWithPanel button open"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel open panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel renders a disabled button 1`] = `
<div
  class="buttonWithPanel button"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      disabled=""
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`shared/ButtonWithPanel renders the ButtonWithPanel with a closed panel 1`] = `
<div
  class="buttonWithPanel button"
>
  <div
    class="buttonWithPanelButtonWrapper"
  >
    <input
      class="buttonWithPanelButton buttonButton"
      type="button"
      value="My Button"
    />
  </div>
  <div
    class="arrowPanelAnchor"
  >
    <div
      class="arrowPanel panel"
    >
      <div
        class="arrowPanelArrow"
      />
      <div
        class="arrowPanelContent"
      >
        <div>
          Panel content
        </div>
      </div>
    </div>
  </div>
</div>
`;
